### PR TITLE
mobile: local authentication

### PR DESCRIPTION
**Pull request by jr-simplex**

### CODE REVIEW

File: apps/ios/Shared/ContentView.swift [hunk 1]
```diff
@@ -5,18 +5,21 @@
...
     @ObservedObject var alertManager = AlertManager.shared
     @ObservedObject var callController = CallController.shared
     @State private var showNotificationAlert = false
+    @State private var showChats = false
 
     var body: some View {
         ZStack {
             if let step = chatModel.onboardingStage {
-                if case .onboardingComplete = step,
+                if showChats,
```
Comment (epoberezkin) [inline apps/ios/Shared/ContentView.swift:21]: the name is confusing, particularly after the call is also here...

File: apps/android/app/src/main/java/chat/simplex/app/views/helpers/LocalAuthentication.kt [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    else -> false
+  }
+}
+
+fun authenticate(
+  promptTitle: String,
+  promptSubtitle: String,
+  activity: FragmentActivity,
+  context: Context,
+  onLAResult: (laResult: LAResult) -> Unit
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/views/helpers/LocalAuthentication.kt:32]:
```suggestion
  completed: (LAResult) -> Unit
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 3]
```diff
@@ -46,20 +46,40 @@
...
   <string name="it_can_disabled_via_settings_notifications_still_shown"><b>It can be disabled via settings</b> – notifications will still be shown while the app is running.</string>
   <string name="turn_off_battery_optimization">In order to use it, please <b>disable battery optimization</b> for <xliff:g id="appName">SimpleX</xliff:g> in the next dialog. Otherwise, the notifications will be disabled.</string>
   <string name="turning_off_background_service">Battery optimization is active, turning off Private notifications. You can re-enable them via settings.</string>
 
+  <!-- local authentication notice - SimpleXAPI.kt -->
+  <string name="la_notice_title">Local authentication</string>
+  <string name="la_notice_text">You can enable local authentication for <xliff:g id="appNameFull">SimpleX Chat</xliff:g> to protect your chats. If you accept, you will be prompted to complete authentication before this feature is enabled.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:65]:
```suggestion
  <string name="la_notice_text">To protect your chats, enable local authentication for <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nYou will be prompted to complete authentication before this feature is enabled.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 4]
```diff
@@ -46,20 +46,40 @@
...
+  <string name="la_notice_turn_on">Turn on</string>
+
+  <!-- LocalAuthentication.kt -->
+  <string name="auth_turned_on">Local authentication turned on</string>
+  <string name="auth_turned_on_desc">You will be required to complete authentication each time you start the app. App will also lock after being in background for more than 30 seconds.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:70]:
```suggestion
  <string name="auth_turned_on_desc">You will be required to complete authentication every time you start or resume the app after more than 30 seconds in background.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -46,20 +46,40 @@
...
+  <string name="auth_access_chats">Access chats</string>
+  <string name="auth_log_in_using_credential">Log in using your credential</string>
+  <string name="auth_enable">Enable local authentication</string>
+  <string name="auth_disable">Disable local authentication</string>
+  <string name="auth_confirm_credential">Confirm your credential</string>
+  <string name="auth_error">Authentication error</string>
+  <string name="auth_error_w_desc">Authentication error: <xliff:g id="desc">%1$s</xliff:g></string>
+  <string name="auth_failed">Authentication failed</string>
+  <string name="auth_unavailable">Authentication unavailable</string>
+  <string name="auth_unavailable_instruction_desc">Your device is not configured for authentication. If your device supports authentication via device credential (e.g. PIN) or biometric authentication, you can set it up and then enable this feature from Settings to protect your chats.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:80]:
```suggestion
  <string name="auth_unavailable_instruction_desc">Your device is not configured for authentication. You can enable authentication later via Settings, once you enable device authentication.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 6]
```diff
@@ -46,20 +46,40 @@
...
+  <string name="auth_unavailable_turning_off_desc">Your device is not configured for authentication. Turning off local authentication for <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:81]:
```suggestion
  <string name="auth_unavailable_turning_off_desc">Device authentication is not enabled. Local authentication for <xliff:g id="appNameFull">SimpleX Chat</xliff:g> is turned off.</string>
```

File: apps/android/app/src/main/java/chat/simplex/app/MainActivity.kt [hunk 7]
```diff
@@ -31,26 +30,29 @@ import chat.simplex.app.views.helpers.*
...
   private val chatController by lazy { (application as SimplexApp).chatController }
+  private val userAuthorized = mutableStateOf<Boolean?>(null)
+  private val lastLA = mutableStateOf<Long?>(null)
 
   override fun onCreate(savedInstanceState: Bundle?) {
     super.onCreate(savedInstanceState)
+    ProcessLifecycleOwner.get().lifecycle.addObserver(this)
 //    testJson()
-    processNotificationIntent(intent, vm.chatModel)
+    val cm = vm.chatModel
```
Comment (epoberezkin) [inline apps/android/app/src/main/java/chat/simplex/app/MainActivity.kt:46]:
```suggestion
    val m = vm.chatModel
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 8]
```diff
@@ -64,6 +64,26 @@
...
     <string name="simplex_service_notification_text">Приём сообщений…</string>
 
+    <!-- local authentication notice - SimpleXAPI.kt -->
+    <string name="la_notice_title">Заблокируйте свои чаты</string>
+    <string name="la_notice_text">Чтобы защитить свои чаты, включите блокировку чатов для <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nВам будет предложено пройти аутентификацию, прежде чем эта функция будет включена.</string>
+    <string name="la_notice_turn_on">Включить</string>
+
+    <!-- LocalAuthentication.kt -->
+    <string name="auth_turned_on">Блокировка чатов включена</string>
+    <string name="auth_turned_on_desc">Вы будете аутентифицированы при каждом запуске приложения и при возобновлении его работы спутся 30 секунд в фоновом режиме.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:74]:
```suggestion
    <string name="auth_turned_on_desc">Вы будете аутентифицированы при каждом запуске и возобновлении приложения после 30 секунд в фоновом режиме.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 9]
```diff
@@ -64,6 +64,26 @@
     <string name="simplex_service_notification_title"><xliff:g id="appNameFull">SimpleX Chat</xliff:g> сервис</string>
     <string name="simplex_service_notification_text">Приём сообщений…</string>
 
+    <!-- local authentication notice - SimpleXAPI.kt -->
+    <string name="la_notice_title">Заблокируйте свои чаты</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:68]:
```suggestion
    <string name="la_notice_title">Блокировка чата</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 10]
```diff
@@ -64,6 +64,26 @@
...
+    <string name="la_notice_text">Чтобы защитить свои чаты, включите блокировку чатов для <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nВам будет предложено пройти аутентификацию, прежде чем эта функция будет включена.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:69]:
```suggestion
    <string name="la_notice_text">Чтобы защитить ваши чаты, включите блокировку чатов для <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nВам будет нужно пройти аутентификацию для включения блокировки.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 11]
```diff
@@ -64,6 +64,26 @@
...
+    <string name="auth_log_in_using_credential">Пройдите аутентификацию</string>
+    <string name="auth_enable">Включить блокировку чатов</string>
+    <string name="auth_disable">Отключить блокировку чатов</string>
+    <string name="auth_confirm_credential">Пройдите аутентификацию</string>
+    <string name="auth_error">Ошибка аутентификации</string>
+    <string name="auth_error_w_desc">Ошибка аутентификации: <xliff:g id="desc">%1$s</xliff:g></string>
+    <string name="auth_failed">Ошибка аутентификации</string>
+    <string name="auth_unavailable">Аутентификация недоступна</string>
+    <string name="auth_unavailable_instruction_desc">На устройстве не включена аутентификация. Вы можете включить блокировку чатов в Настройках после включения аутентификации.</string>
+    <string name="auth_unavailable_turning_off_desc">На устройстве выключена аутентификация. Отключение блокировки чатов.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:85]:
```suggestion
    <string name="auth_unavailable_turning_off_desc">На устройстве выключена аутентификация. Отключение блокировки SimpleX Chat.</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 12]
```diff
@@ -64,6 +64,26 @@
...
+    <string name="auth_access_chats">Разблокировать чаты</string>
+    <string name="auth_log_in_using_credential">Пройдите аутентификацию</string>
+    <string name="auth_enable">Включить блокировку чатов</string>
+    <string name="auth_disable">Отключить блокировку чатов</string>
+    <string name="auth_confirm_credential">Пройдите аутентификацию</string>
+    <string name="auth_error">Ошибка аутентификации</string>
+    <string name="auth_error_w_desc">Ошибка аутентификации: <xliff:g id="desc">%1$s</xliff:g></string>
+    <string name="auth_failed">Ошибка аутентификации</string>
+    <string name="auth_unavailable">Аутентификация недоступна</string>
+    <string name="auth_unavailable_instruction_desc">На устройстве не включена аутентификация. Вы можете включить блокировку чатов в Настройках после включения аутентификации.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:84]:
```suggestion
    <string name="auth_unavailable_instruction_desc">На устройстве не включена аутентификация. Вы можете включить блокировку чата в Настройках после включения аутентификации.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -64,6 +64,26 @@
   <string name="simplex_service_notification_title"><xliff:g id="appNameFull">SimpleX Chat</xliff:g> service</string>
   <string name="simplex_service_notification_text">Receiving messages…</string>
 
+  <!-- local authentication notice - SimpleXAPI.kt -->
+  <string name="la_notice_title">Lock your chats</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:68]:
```suggestion
  <string name="la_notice_title">Lock your chat</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 14]
```diff
@@ -46,13 +46,13 @@
...
     <string name="cannot_delete_contact">Невозможно удалить контакт!</string>
     <string name="contact_cannot_be_deleted_as_they_are_in_groups">Контакт <xliff:g id="contactName" example="Jane Doe">%1$s!</xliff:g> не может быть удален, так как является членом групп(ы) <xliff:g id="groups" example="[team, chess club]">%2$s</xliff:g>.</string>
-    <string name="icon_descr_instant_notifications">Мгновенные уведомления</string>
     <string name="error_deleting_contact">Ошибка удаления контакта</string>
     <string name="error_deleting_group">Ошибка удаления группы</string>
     <string name="error_deleting_contact_request">Ошибка удаления запроса</string>
     <string name="error_deleting_pending_contact_connection">Ошибка удаления ожидаемого соединения</string>
 
     <!-- background service notice - SimpleXAPI.kt -->
+    <string name="icon_descr_instant_notifications">Мгновенные уведомления</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:55]: why not just use the next one?

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 15]
```diff
@@ -64,6 +64,26 @@
     <string name="simplex_service_notification_title"><xliff:g id="appNameFull">SimpleX Chat</xliff:g> сервис</string>
     <string name="simplex_service_notification_text">Приём сообщений…</string>
 
+    <!-- local authentication notice - SimpleXAPI.kt -->
+    <string name="la_notice_title">Заблокируйте свои чаты</string>
+    <string name="la_notice_text">Чтобы защитить свои чаты, включите блокировку чатов для <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nВам будет предложено пройти аутентификацию, прежде чем эта функция будет включена.</string>
+    <string name="la_notice_turn_on">Включить</string>
+
+    <!-- LocalAuthentication.kt -->
+    <string name="auth_turned_on">Блокировка чатов включена</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:73]:
```suggestion
    <string name="auth_turned_on">Блокировка SimpleX включена</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 16]
```diff
@@ -64,6 +64,26 @@
...
+    <string name="auth_turned_on_desc">Вы будете аутентифицированы при каждом запуске приложения и при возобновлении его работы спутся 30 секунд в фоновом режиме.</string>
+    <string name="auth_access_chats">Разблокировать чаты</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:75]:
```suggestion
    <string name="auth_access_chats">Разблокировать SimpleX</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml [hunk 17]
```diff
@@ -64,6 +64,26 @@
...
+    <string name="auth_log_in_using_credential">Пройдите аутентификацию</string>
+    <string name="auth_enable">Включить блокировку чатов</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values-ru/strings.xml:77]:
```suggestion
    <string name="auth_enable">Включить блокировку SimpleX</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="la_notice_text">To protect your chats, turn on Chat Lock for <xliff:g id="appNameFull">SimpleX Chat</xliff:g>.\nYou will be prompted to complete authentication before this feature is enabled.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:69]:
```suggestion
  <string name="la_notice_text">To protect your information, turn on SimpleX Lock.\nYou will be prompted to complete authentication before this feature is enabled.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="la_notice_turn_on">Turn on</string>
+
+  <!-- LocalAuthentication.kt -->
+  <string name="auth_turned_on">Chat Lock turned on</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:73]:
```suggestion
  <string name="auth_turned_on">SimpleX Lock turned on</string>
```
```suggestion
  <string name="auth_turned_on">Chat Lock turned on</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="auth_turned_on_desc">You will be required to complete authentication every time you start or resume the app after more than 30 seconds in background.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:74]:
```suggestion
  <string name="auth_turned_on_desc">You will be required to authenticate when you start or resume the app after 30 seconds in background.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 21]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="auth_access_chats">Access chats</string>
+  <string name="auth_log_in_using_credential">Log in using your credential</string>
+  <string name="auth_enable">Enable Chat Lock</string>
+  <string name="auth_disable">Disable Chat Lock</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:78]:
```suggestion
  <string name="auth_enable">Enable SimpleX Lock</string>
  <string name="auth_disable">Disable SimpleX Lock</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 22]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="auth_confirm_credential">Confirm your credential</string>
+  <string name="auth_error">Authentication error</string>
+  <string name="auth_error_w_desc">Authentication error: <xliff:g id="desc">%1$s</xliff:g></string>
+  <string name="auth_failed">Authentication failed</string>
+  <string name="auth_unavailable">Authentication unavailable</string>
+  <string name="auth_unavailable_instruction_desc">Device authentication is not enabled. You can turn on Chat Lock via Settings, once you enable device authentication.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:84]:
```suggestion
  <string name="auth_unavailable_instruction_desc">Device authentication is not enabled. You can turn on SimpleX Lock via Settings, once you enable device authentication.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 23]
```diff
@@ -64,6 +64,26 @@
...
+  <string name="auth_unavailable_turning_off_desc">Device authentication is disabled. Turning off Chat Lock.</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:85]:
```suggestion
  <string name="auth_unavailable_turning_off_desc">Device authentication is disabled. Turning off SimpleX Lock.</string>
```

File: apps/android/app/src/main/res/values/strings.xml [hunk 24]
```diff
@@ -236,6 +256,7 @@
   <string name="chat_with_the_founder">Connect to the developers</string>
   <string name="send_us_an_email">Send us email</string>
   <string name="private_notifications">Private notifications</string>
+  <string name="chat_lock">Chat Lock</string>
```
Comment (epoberezkin) [inline apps/android/app/src/main/res/values/strings.xml:259]:
```suggestion
  <string name="chat_lock">SimpleX Lock</string>
```

File: apps/android/app/src/main/res/values-ru/strings.xml (hunk 14 above)
Comment (jr-simplex) [reply apps/android/app/src/main/res/values-ru/strings.xml:55]: actually I'm not sure why it's in diff but let's keep it

**Pull request merged**